[Feature] Snackbar – Update default offset to assume nav bar

narin opened the pull request:
<!-- PR title naming convention:
'[Issue type] Brief summary of issue suitable for changelog or copy/paste issue title',
where Issue type = bug, feature, spike, CU (code upkeep), etc.-->

<!-- Preferred branch naming convention:
'[Issue type]/[Issue #]-[Your name]-[Summary of issue]',
where Issue type = bug, feature, spike, CU (code upkeep), etc.-->

## Description of Change
<!-- Describe the change and context with which it was made beyond ACs unless straightforward.
Consider:
 - What is relevant to code reviewer(s) and not in the ticket?
 - What context may be relevant to a future dev or you in 6 months about this PR?
 - Did the course of work lead to notable dead ends? If so, why didn't they pan out?
 - Did the change add new dependencies? Why?
 - Were there important sources to link? Examples: an open bug with a dependency project, an article of someone else solving the same problem that was partially or wholly copied, external documentation relevant to solution
 -->
Updates default offset to match app and assume there is a nav bar present. Values copied from [va-mobile-app's standardTheme.ts](https://github.com/department-of-veterans-affairs/va-mobile-app/blob/3fd975d714b3ed4628353b343270f1d0c6291c3c/VAMobile/src/styles/themes/standardTheme.ts#L301).

Also adds util to check for iOS devices with home buttons and adjust offset depending on that. 

Default offset will eventually be replaced by a global setting that can be adjusted.

In regards to the safe areas, based on my research, nesting multiple <SafeAreaProvider> or using <SafeAreaView> within <SafeAreaProvider>s is not recommended. The mobile app already has a SafeAreaProvider so we won't include it here but made note of it in the docs.

#### Testing Packages
<!-- List or range of alpha/beta packages published in association with this PR, if any -->
- 0.21.1-alpha.11
- 0.21.1-alpha.12
- 0.21.1-alpha.13
- 0.21.1-alpha.15

### Screenshots/Video
<!-- Add screenshots or video as needed; before/after recommended if appropriate. 
Convenience side-by-side formatting:
Before/after: <img src="" width="49%" />&nbsp;&nbsp;<img src="" width="49%" />
Accordion before/after: <details><summary>Before/after</summary><img src="" width="49%" />&nbsp;&nbsp;<img src="" width="49%" /></details>
-->
<details><summary>DS snackbar vs App Snackbar (iOS)</summary><img src="https://github.com/user-attachments/assets/f7a18c3d-9a00-47cd-9fc2-e464357a6aab" width="49%"/>
<img src="https://github.com/user-attachments/assets/15774f53-05a1-4615-bff8-93067cdcfd68" width="49%"/></details>


<details><summary>DS snackbar vs App Snackbar (iPhone SE with home button)</summary><img src="https://github.com/user-attachments/assets/a3d7b21d-931e-46a5-b3d8-cddec2b74385" width="49%"/>
<img src="https://github.com/user-attachments/assets/bd9e47e4-dfff-43d6-bdfe-8b614e47a777" width="49%"/></details>


<details><summary>DS snackbar vs App Snackbar (Android)</summary><img src="https://github.com/user-attachments/assets/20794445-11b7-4025-949d-e4c23b3e1261" width="49%"/><img src="https://github.com/user-attachments/assets/f5977792-b3f4-4269-a85e-85a3d194ee84" width="49%"/></details>



## Testing
<!-- Describe testing conducted to validate changes.
Consider highlighting:
- What testing was not explicitly done and may be relevant for QA? 
- Edge cases validated
- Special situations that could not be tested
- Any testing performed in a consuming app -->

- [x] Tested on iOS <!-- simulator is fine -->
- [x] Tested on Android <!-- simulator is fine -->
- [x] Tested on Web

## PR Checklist
Code reviewer validation:
- General
	- [x] PR is linked to ticket(s)
	- [x] PR has `changelog` label applied if it's to be included in the changelog
	- [x] Acceptance criteria: 
		- All satisfied _or_
		- Documented reason for not being performed _or_
		- Split to separate ticket and ticket is linked by relevant AC(s)
	- [x] Above PR sections adequately filled out
	- [x] If any breaking changes, in accordance with the [pre-1.0.0 versioning guidelines](https://github.com/department-of-veterans-affairs/va-mobile-library#versioning-policy): a CU ticket has been created for the VA Mobile App detailing necessary adjustments with the package version that will be published by this ticket
- Code
	- [x] Tests are included if appropriate (or split to separate ticket)
	- [x] New functions have proper TSDoc annotations

## Publish
<!-- Most changes entail a version increment; section can be removed for PRs exclusively within non-ship-relevant files (e.g. unit tests, Storybook stories) -->
If changes warrant a new version [per the versioning guidelines](https://github.com/department-of-veterans-affairs/va-mobile-library#versioning-policy) and the PR is approved and ready to merge:
Doesn't warrant publish.  Will publish after #440 since that should be the last ticket with functional changes.

TimRoe commented on packages/components/src/utils/OSfunctions.tsx:
```diff
@@ -110,3 +110,29 @@ export function useExternalLink(): (
+export const hasHomeButton = () => {
+  const { height, width } = Dimensions.get('window')
+
+  if (Platform.OS !== 'ios') {
```
```suggestion
  if (!isIOS()) {
```

TimRoe replied:
Nitpicky: could also move above the Dimensions call since it just gets dropped on the floor

TimRoe commented on packages/components/src/utils/OSfunctions.tsx:
```diff
@@ -110,3 +110,29 @@ export function useExternalLink(): (
+    return false
+  }
+
+  const devicesWithHomeButton = [
+    { height: 667, width: 375 }, // iPhone 6, 6s, 7, 8, SE (2nd gen)
+    { height: 736, width: 414 }, // iPhone 6+, 6s+, 7+, 8+
+    { height: 568, width: 320 }, // iPhone 5, 5s, 5c, SE (1st gen)
+  ]
+
+  const currentDevice = { height, width }
+
+  return devicesWithHomeButton.some(
+    (device) =>
+      device.height === currentDevice.height &&
+      device.width === currentDevice.width,
+  )
```
I'm not sure about this. It feels brittle like new devices (e.g. iPhone SE 4th gen) could have a new resolution and things could break on us--especially since the list doesn't note the SE (3rd gen), but looks like it has the same resolution as the 2nd gen.

Generally: not sure how I feel about us being in the business of manually handling this.

narin replied:
Do you have an alternative suggestion? iPhone SE 4th gen [is rumored](https://www.macrumors.com/guide/iphone-se-4/) to no longer have a home button so this list shouldn't change.

TimRoe replied:
We couldn't do native module stuff at all without converting the project, right? Had found [this](https://stackoverflow.com/questions/52652956/how-to-detect-if-the-device-iphone-has-physical-home-button) on basically querying the insets without directly using them.

Otherwise I guess not if not leaning on `react-native-safe-area-context` in some capacity, whether directly or using as a means to do a similar query to the above without native.

TimRoe commented on packages/components/src/utils/OSfunctions.tsx:
```diff
@@ -110,3 +110,29 @@ export function useExternalLink(): (
+/**
+ * Returns whether or not the device has a home button
+ */
+export const hasHomeButton = () => {
```
Should be capitalized?

TimRoe reviewed (approved):
Approved with some outstanding nitpicky things. Think the only notable think is a separate ticket to add unit tests  for the new function so we don't forget.

narin commented:
Since we know the app is using `react-native-safe-area-context` decided to leverage that instead and provider a helper provider for apps not using it.
- Created a new hook `useSnackbarDefaultOffset()` which uses `useSafeAreaInsets()` bottom inset to determine default offset
- Fixed Android offset so it's the same as iOS
- Add `<SnackbarWithSafeArea/>` convenience provider for apps that don't have a `<SafeAreaProvider/>`
- Remove `hasHomeButton()` util

I assume by the time another app is using our lib, we will have figured out native code by then and could switch to using `react-native-device-info` or similar libs to check for home buttons.

TimRoe commented on packages/components/src/components/Snackbar/Snackbar.tsx:
```diff
@@ -90,17 +82,17 @@ export type SnackbarProps = Omit<ToastProps, 'data' | 'message'> &
- * offset from the bottom of the screen may not display as expected.
+ * **Note:** The snackbar requires [react-native-safe-area-context](https://github.com/th3rdwave/react-native-safe-area-context).
+ * SnackbarProvider should be placed within SafeAreaProvider. If you do not have a SafeAreaProvider, you can use
+ * SnackbarWithSafeAreaProvider instead.
```
```suggestion
 * SnackbarProviderWithSafeArea instead.
```

TimRoe commented on packages/components/src/components/Snackbar/SnackbarProvider.tsx:
```diff
@@ -1,7 +1,9 @@
 import React, { ReactNode, createContext, useState } from 'react'
 
-import { SNACKBAR_DEFAULT_OFFSET, Snackbar } from './Snackbar'
+import { SafeAreaProvider } from 'react-native-safe-area-context'
+import { Snackbar, SnackbarProps } from './Snackbar'
 import { ToastProvider } from 'react-native-toast-notifications'
```
Nitpicky: mixing library and internal file imports

TimRoe commented on packages/components/src/components/Snackbar/useSnackbarDefaultOffset.tsx:
```diff
@@ -0,0 +1,17 @@
+    )
+  }
+
+  const { bottom } = insets
+  return bottom === 0 ? 60 : 60 + bottom
```
```suggestion
  return 60 + insets.bottom
```

TimRoe replied:
May also be worth making 60 a const or having a comment noting it represents the height of the Nav bar

TimRoe commented on packages/components/src/components/Snackbar/Snackbar.tsx:
````diff
@@ -86,12 +82,18 @@ export type SnackbarProps = Omit<ToastProps, 'data' | 'message'> &
  * )
  * ```
  *
+ * **Note:** The snackbar requires [react-native-safe-area-context](https://github.com/th3rdwave/react-native-safe-area-context).
````
Need to move `react-native-safe-area-context` from a devDependency to dependency for it to work in consuming apps that wouldn't otherwise have it.

TimRoe reviewed (changes requested):
Good revisions, makes much more sense in my mind.

Marking request changes because we need to change `react-native-safe-area-context` to a dependency or `SnackbarProviderWithSafeArea` wouldn't actually work within a consuming app without them also adding the package.

Beyond that:
- Some minor nitpick suggestions here and there
- Don't feel strongly, but wonder if `useSnackbar` and `useSnackbarDefaultOffset` instead of being their own files should be together in a `hooks.tsx` or something (still within the Snackbar folder)

narin commented on packages/components/src/components/Snackbar/Snackbar.tsx:
```diff
@@ -165,6 +172,10 @@ export const Snackbar: FC<SnackbarProps> = (toast) => {
     },
   }
 
+  if (isAndroid()) {
+    containerProps.tabIndex = 0
+  }
+
```
Had to move this out because in the app, running on iOS would get the following error:

`node_modules/@department-of-veterans-affairs/mobile-component-library/src/components/Snackbar/Snackbar.tsx(160,5): error TS2353: Object literal may only specify known properties, and 'tabIndex' does not exist in type 'ViewProps'.`

TimRoe replied:
Weird. I had noticed TS giving a red squiggly line when prototyping in the app node_modules on one ticket, but it seemed to run fine in iOS/Android previously. Oh well.

narin commented:
- Addressed nits
- Putting `react-native-safe-context` as a dependency caused a conflict and crash within the app. Put it in `peerDependencies` instead
- Added `isAndroid()` check to the `tabIndex` for the Snackbar container props to avoid TS error when running in app on iOS
- Bumped mobile-tokens in peerDependencies since the new color token from 0.14.0 is needed in Snackbar
- Keeping hooks separated. I think I tried this initially but was getting a circular dependency warning

TimRoe commented on packages/components/src/components/Snackbar/useSnackbar.tsx:
```diff
@@ -20,11 +20,9 @@ export function useSnackbar() {
-    const { offset, setOffset } = context
+  const { offset, setOffset } = context
+
+  const show = (message: string, snackbarOptions?: SnackbarOptions) => {
```
Doesn't matter, but just noting that `const { offset, setOffset } = context` was moved out of `show`. Not sure if intentional or just a minor inconsistency fixing conflicts with my changes.

narin replied:
Will be cleaned up further in the screenreader dismiss ticket #440

TimRoe commented on packages/components/src/components/Snackbar/Snackbar.tsx:
```diff
@@ -181,7 +158,6 @@ export const Snackbar: FC<SnackbarProps> = (toast) => {
   const containerProps: ViewProps = {
     accessibilityViewIsModal: true, // iOS only
-    tabIndex: 0, // Android only
     // Above props prevent screen reader from tap focusing elements behind the Snackbar
```
Think we need to duplicate this comment or otherwise rework slightly for clarity--maybe pull the iOS one out as well for something like:
```
  // Below props prevent screen reader from tap focusing elements behind the Snackbar
  if (isAndroid()) {
    containerProps.tabIndex = 0
  }
  if (isIOS()) {
    containerProps.accessibilityViewIsModal = true
  }
```
?

TimRoe commented on packages/components/src/components/Snackbar/SnackbarProvider.tsx:
```diff
@@ -1,7 +1,9 @@
+import { ToastProvider } from 'react-native-toast-notifications'
 import React, { ReactNode, createContext, useState } from 'react'
```
```suggestion
import { SafeAreaProvider } from 'react-native-safe-area-context'
import { ToastProvider } from 'react-native-toast-notifications'
import React, { ReactNode, createContext, useState } from 'react'
```

TimRoe commented on packages/components/src/components/Snackbar/SnackbarProvider.tsx:
```diff
@@ -1,7 +1,9 @@
 
-import { SNACKBAR_DEFAULT_OFFSET, Snackbar, SnackbarProps } from './Snackbar'
-import { ToastProvider } from 'react-native-toast-notifications'
+import { SafeAreaProvider } from 'react-native-safe-area-context'
```
```suggestion

```
with above suggestion

TimRoe commented on packages/components/src/components/Snackbar/SnackbarProvider.tsx:
```diff
@@ -24,10 +28,18 @@ export const SnackbarProvider: React.FC<{ children: ReactNode }> = ({
         offset={offset}
         placement="bottom"
-        renderToast={(toast) => <Snackbar {...toast as SnackbarProps} />}
+        renderToast={(toast) => <Snackbar {...(toast as SnackbarProps)} />}
```
```suggestion
        renderToast={(toast) => <Snackbar {...toast as SnackbarProps} />}
```
?

narin replied:
Autoformatter forces this for some reason. I can save without formatting

TimRoe reviewed (approved):
Approved w/ some final minor nitpicks.
